Initial info tree and material=>recycling info conversion object

adamosimpson21 opened the pull request:
nodeList and whereDoesItGo files only contain Array of objects used in infoTreeLogic page. That page converts one of these objects (currently the one housed in whereDoesItGo) and converts it into a tree structure, with child=>parent and parent=> child relationships. I've left comments in the code describing the specific functions and progress that can be made in the future

adamosimpson21 committed:
Initial info tree and material=>recycling info conversion object

nodeList and whereDoesItGo files only contain Array of objects used in infoTreeLogic page. That page converts one of these objects (currently the one housed in whereDoesItGo) and converts it into a tree structure, with child=>parent and parent=> child relationships. I've left comments in the code describing the specific functions and progress that can be made in the future

AndyCErnst commented on infoTree/nodeList.js:
```diff
@@ -0,0 +1,3 @@
+//I believe this is the list of every branch with category titles, but it's too late for me to really look at it right now.
```
The "it's too late for me to really look at it right now." seems like it should be a PR comment, not a code comment

AndyCErnst replied:
Also, you don't mention that this list is from seattle.gov. Right now I wouldn't know if this is the raw or cleaned data.

AndyCErnst commented on infoTree/recycleBuddyScratch.js:
```diff
@@ -0,0 +1,200 @@
+//import the current list of leaves from separate file
```
comments describing how javascript works aren't useful. I can see where you might want more of this kind of comment in a project like this, but `import` is easy enough to look up if people don't understand, and it's already used in most files.

AndyCErnst commented on infoTree/recycleBuddyScratch.js:
```diff
@@ -0,0 +1,200 @@
+import whereDoesItGo from './whereDoesItGo';
+import nodeList from './nodeList'
+
+//I found this on a github repo (under one of my starred Repos) so obviously it's not my own work
```
link to the repo, give credit where it's due. I'd shorten the comment to something like:
`Tree implementation originally from <repo>. Trimmed down from the original version.`

adamosimpson21 replied:
Cool. I didn't really know how to handle this and discuss it through comments. That's much better than what I did. will change

AndyCErnst commented on infoTree/recycleBuddyScratch.js:
```diff
@@ -0,0 +1,200 @@
+}
+
+
+//formatting whereRecycleProto. Changes the value string to an array. 
```
This looks like where your work starts. I'd like to see a clean separation between code that we don't own and code that we do. If we are heavily modifying this we might consider it _ours_ at some point, but for now I'd much rather see it as a library installed with npm, or in a vendor folder. If you are modifying it, I want to see your modifications all in the same file so it's obvious what's going on.

AndyCErnst commented on infoTree/recycleBuddyScratch.js:
```diff
@@ -0,0 +1,200 @@
+function formatRecycleObj(whereRecycleObj){
+  for(let i = 0; i<whereRecycleObj.length; i++){
+  	//splits into array and slices from WhereDoesItGo/[......]/index.htm
+    whereRecycleObj[i].value = whereRecycleObj[i].value.split("/").slice(4, -1)
```
clean code: avoid traditional for loops. This would be clearer in a forEach.

AndyCErnst commented on infoTree/recycleBuddyScratch.js:
```diff
@@ -0,0 +1,200 @@
+function createRecycleTree(recycleArray){
+  //format object so its easier to use if it's not formatted
+  if(typeof recycleArray[0].value==="string"){
+    recycleArray = formatRecycleObj(recycleArray);
```
nice that this data cleaner is its own function.

AndyCErnst commented on infoTree/recycleBuddyScratch.js:
```diff
@@ -0,0 +1,200 @@
+      if(!recycleTree.contains(recycleArray[i].value[j], recycleTree.traverseBF)){
+        // console.log(recycleArray[i].value[j])
+        // checks to make sure the branch's parent isn't the rootNode
+        if(recycleArray[i].value[j-1]){
```
forEach loops will clear up a lot of `i`s and `j`s. The value of `recycleArray[i].value[j-1]` is particularly hard to understand, even with your comment

AndyCErnst commented on infoTree/whereDoesItGo.js:
```diff
@@ -0,0 +1,826 @@
+//this list only contains the leaves of the tree structure
```
be more explicit, "These are the leaves of a tree generated from the Seattle.gov recycling website using recycleBuddyScratch.js"

AndyCErnst replied:
Although in the end, I don't imagine we are keeping this file. This is just to feed the crawler right?

AndyCErnst commented on infoTree/recycleBuddyScratch.js:
```diff
@@ -0,0 +1,200 @@
+}
+
+//call the function
+createRecycleTree(whereDoesItGo);
```
seems silly to call the function then throw away the result

adamosimpson21 replied:
yea, I was mostly using it to view what I had in the console, as we're not using it anywhere yet. Eventually It'll be written to a file or imported in some way

AndyCErnst commented on infoTree/recycleBuddyScratch.js:
```diff
@@ -0,0 +1,200 @@
+    //for every branch, this creates the branch structure, but no the leaves
+    for(let j=0; j<recycleArray[i].value.length; j++){
+      //check to see if the branch is in the tree already
+      if(!recycleTree.contains(recycleArray[i].value[j], recycleTree.traverseBF)){
```
their `contains` doesn't take a second parameter.

AndyCErnst reviewed (changes requested):
Eventually, we should delete the data files that won't go in the final version. We should at least put them in a data folder so it's clear they aren't code.

Nice job tackling this. This is probably the most data structure heavy part of the app we'll need to build

adamosimpson21 committed:
Refactored createRecycleTree and Added data flow structure

Renamed recycleBuddyScratch to createRecycleTree. Refactored createRecycleTree to use forEach loops and clean up a lot of other aspects of the code. Increased quality of comments. SeattleGovScrape serves as a prototype space for scrapeHtml function. createDataStore demonstrates how data will flow. This file is mostly a skeleton, with many TODOs and empty functions

AndyCErnst commented on infoTree/createDataStore.js:
```diff
@@ -0,0 +1,32 @@
+function scrapeHtml(html){
+	var howToDispose = {};
+	//Input: HTML from seattle.gov where does it go site
+	//Returns an object that holds formatted data as described in seattleGovScrape.js
```
comments like this are more typically in javadoc style, something like
```
 /**
 * Simplifies and normalizes scraped html into a usable datastructure.
 * 
 * @param html   scraped html from Seattle.gov
 * @return tree of html nodes 
 */
```

adamosimpson21 replied:
sweet! Thanks for showing me this

AndyCErnst commented on infoTree/createRecycleTree.js:
```diff
@@ -0,0 +1,208 @@
+
+//finds the URL to scrape data from by adding parent's name to base URL
+Tree.prototype.findUrl = function(leaf){
+	url = "https://www.seattle.gov/util/MyServices/WhereDoesItGo/"
```
no variable declaration for `url`, this makes it a global which we definitely want to avoid.

AndyCErnst commented on infoTree/createRecycleTree.js:
```diff
@@ -0,0 +1,208 @@
+//This could be done with regular expressions, but an array containing the parents seems easier to me.
+//TODO: adapt for different inputs. because of the slice(4,-1) this could break easily when we include the category titles
+function formatRecycleArray(whereRecycleArray){
+  whereRecycleArray.forEach(branchInfo => branchInfo.value = branchInfo.value.split("/").slice(4,-1))
```
I'd break `branchInfo.value.split("/").slice(4,-1)` into its own function with a descriptive name

AndyCErnst commented on infoTree/createRecycleTree.js:
```diff
@@ -0,0 +1,208 @@
+  recycleTree.add("rootNode")
+
+  //format object so its easier to use if it's not formatted
+  (typeof recycleArray[0].value==="string") && (recycleArray = formatRecycleArray(recycleArray))
```
what is going on here? It seems like this wants to be 
```
if (typeof recycleArray[0].value==="string") {
   recycleArray = formatRecycleArray(recycleArray);
}
```

adamosimpson21 replied:
I was just trying to be fancy. Probably too fancy

AndyCErnst replied:
lol, read this https://hackernoon.com/why-senior-devs-write-dumb-code-and-how-to-spot-a-junior-from-a-mile-away-27fa263b101a

AndyCErnst commented on infoTree/createRecycleTree.js:
```diff
@@ -0,0 +1,208 @@
+
+  recycleArray.forEach(branchInfo => {
+    branchInfo.value.forEach((branchName, index) => 
+      !recycleTree.contains(branchName) && 
```
this will be easier to read if you pull it out to a filter
```
branchInfo.value.filter( branchName => recycleTree.contains(branchName)).forEach((branchName, index) =>  
```

adamosimpson21 replied:
I'm having trouble with filter because the way I'm doing it, I think it initially filters before adding any of the branch elements, so it can't filter out what doesn't exist yet in the tree. I'm going to move the parts around and see if I can come up with anything

AndyCErnst replied:
ah, I missed that this is a dedupe. What happens if you try to add a duplicate? Does it overwrite it? When would there be duplicates?

adamosimpson21 replied:
If you add a duplicate, it makes another node in the tree. We could add an internal structure in the tree to protect against duplicate branches, but It's possible you run into an issue with similar terms (cans, bottles, box, etc. can all be made of different materials). There are many duplicates since you're adding the 2nd level branches each time the forEach runs over them. And you can't exclude them, since the leaf is the information you're looking for

AndyCErnst replied:
In the final decision tree, it would make sense for there to have nodes of  "plastic" with child "box" and "cardboard" with child "box".
I haven't run your code though, I hope to have some time to take a closer look

AndyCErnst commented on infoTree/createRecycleTree.js:
```diff
@@ -0,0 +1,208 @@
+  recycleArray.forEach(branchInfo => {
+    branchInfo.value.forEach((branchName, index) => 
+      !recycleTree.contains(branchName) && 
+        (branchInfo.value[index-1] ? recycleTree.add(branchName, branchInfo.value[index-1]) : recycleTree.add(branchName, "rootNode")
```
I'd separate this out a bit, something like
```
var infoToAdd = branchInfo.value[index-1] || "rootNode";
recycleTree.add(branchName, infoToAdd);
```

adamosimpson21 replied:
that's a great refactor :) thanks

AndyCErnst commented on infoTree/createRecycleTree.js:
```diff
@@ -0,0 +1,208 @@
+//formatting whereDoesItGo. Changes the value string => array. 
+//This could be done with regular expressions, but an array containing the parents seems easier to me.
+//TODO: adapt for different inputs. because of the slice(4,-1) this could break easily when we include the category titles
+function formatRecycleArray(whereRecycleArray){
```
This uses the tree structure and is not part of it, so I would put it in a different file.

AndyCErnst commented on infoTree/createDataStore.js:
```diff
@@ -0,0 +1,32 @@
+
+function scrapeSeattle(url){
+	//TODO: Figure out best methods for getting HTML from a page, probably have to fake some headers
+	// $.(get) ?
```
$.get is a jQuery thing, fetch should have all the options we need

AndyCErnst reviewed (changes requested):
A few code cleanliness things. It's looking good.
You should create some folders for these, maybe a data folder or an output folder.

adamosimpson21 committed:
Initial info tree and material=>recycling info conversion object

nodeList and whereDoesItGo files only contain Array of objects used in infoTreeLogic page. That page converts one of these objects (currently the one housed in whereDoesItGo) and converts it into a tree structure, with child=>parent and parent=> child relationships. I've left comments in the code describing the specific functions and progress that can be made in the future

adamosimpson21 committed:
Serialize Tree

Removed parent component so tree did not loop. Serializable through JSON.stringify()

adamosimpson21 committed:
Serialize Tree

Removed parent component so tree did not loop. Serializable through JSON.stringify()

AndyCErnst commented on infoTree/scrape.html:
```diff
@@ -0,0 +1,13 @@
+  	
+  	<!-- <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script> -->
+  	<script src="./seattleGovScrape.js"></script>
+  </body>
```
why is this file needed? You should be able to run the script in the console.

AndyCErnst commented on infoTree/seattleGovScrape.js:
```diff
@@ -52,73 +57,37 @@ var sampleObject= {
+		key:"AirConditionersHeatPumps",
+		title:"h1",
+	} 
+	var recycleInfo = $('#region').scrape(frame);
```
this API is interesting, I'll have to try this out. The jQuery-like syntax is nice but the scrape method is less straightforward.

AndyCErnst reviewed (commented):
Ideally, this scraper works without a webpage, it's just a command that a person or a build pipeline would run and save the output to a file. The output files should be .json as they are json. That does mean they can't have comments though.

adamosimpson21 committed:
Serialize Tree

updates to PR

adamosimpson21 committed:
Change file type to recycleTree, minor code cleanliness, delete html scratch pad

adamosimpson21 committed:
Merge branch 'infoTree' of https://github.com/Recycle-Buddy/front-end into infoTree

adamosimpson21 committed:
Change file type to recycleTree, minor code cleanliness, delete html scratch pad

adamosimpson21 committed:
Update gitignore

remove *

adamosimpson21 committed:
Delete front-end.iml

adamosimpson21 committed:
Delete misc.xml

adamosimpson21 committed:
Delete modules.xml

adamosimpson21 committed:
Delete vcs.xml

adamosimpson21 committed:
Delete workspace.xml
